### ISSUE TITLE
Question about the authorization flow from JsBffSample with custom IdentityServer 6

### ISSUE THREAD

**Issue by SethNafet** [DuendeSoftware/IdentityServer#938]
When I run the JsBffSample with BackendApiHost enabled in the FrontendHost using demo.duendesoftware.com as Idp, the roundtripping works perfect. However, when setting up my custom Idp using duende.IdentityServer, I cant access the remote API. I always get a 401. 

My BackendApiHost configuration: 

```   
            services.AddControllers();

            services.AddAuthentication("token")
                .AddJwtBearer("token", options =>
                {
                    //options.Authority = "https://demo.duendesoftware.com";
                    //options.Audience = "api";

                    //options.MapInboundClaims = false;
                    options.Authority = "https://localhost:5001";
                    //options.Audience = "api";

                    options.MapInboundClaims = false;
                });

            services.AddAuthorization(options =>
            {
                options.AddPolicy("ApiCaller", policy =>
                {
                    policy.RequireClaim("scope", "api");
                });

                options.AddPolicy("RequireInteractiveUser", policy =>
                {
                    policy.RequireClaim("sub");
                });
            });
```

my FrontendHost configuration:
```          services.AddControllers();

            services.AddBff()
                .AddRemoteApis();

            // registers HTTP client that uses the managed user access token
            services.AddUserAccessTokenHttpClient("api_client", configureClient: client =>
            {
                //client.BaseAddress = new Uri("https://localhost:5002/");
                client.BaseAddress = new Uri("https://localhost:5010/");
            });

            services.AddAuthentication(options =>
            {
                options.DefaultScheme = "cookie";
                options.DefaultChallengeScheme = "oidc";
                options.DefaultSignOutScheme = "oidc";
            })
                .AddCookie("cookie", options =>
                {
                    options.Cookie.Name = "__Host-bff";
                    options.Cookie.SameSite = SameSiteMode.Strict;
                })
                .AddOpenIdConnect("oidc", options =>
                {
                    //options.Authority = "https://demo.duendesoftware.com";
                    options.Authority = "https://localhost:5001";
                    options.ClientId = "interactive.confidential";
                    options.ClientSecret = "secret";
                    options.ResponseType = "code";
                    options.ResponseMode = "query";

                    options.GetClaimsFromUserInfoEndpoint = true;
                    options.MapInboundClaims = false; // set from false
                    options.SaveTokens = true;

                    options.Scope.Clear();
                    options.Scope.Add("openid");
                    options.Scope.Add("profile");
                    options.Scope.Add("email");
                    //options.Scope.Add("customRole");
                    options.Scope.Add("api");
                    options.Scope.Add("offline_access");

                    options.TokenValidationParameters = new()
                    {
                        NameClaimType = "name",
                        RoleClaimType = "role"
                    };
                });
```

My IdentityServer configuration: 
``` 
       private static List<string> AllIdentityScopes =>
            IdentityResources.Select(s => s.Name).ToList();

        private static List<string> AllApiScopes =>
            ApiScopes.Select(s => s.Name).ToList();

        private static List<string> AllScopes =>
            AllApiScopes.Concat(AllIdentityScopes).ToList();

        public static IEnumerable<IdentityResource> IdentityResources =>
            new IdentityResource[]
            {
                new IdentityResources.OpenId(),
                new IdentityResources.Profile(),
                new IdentityResources.Email(),
            };

        public static IEnumerable<ApiScope> ApiScopes =>
            new ApiScope[]
            {
                new ApiScope("customapi"),
                new ApiScope("api")
            };

        public static IEnumerable<ApiResource> ApiResources =>
            new ApiResource[]
            {
                new ApiResource("api", "Demo API", new[] { "name", "email" })
                {
                    ApiSecrets = { new Secret("secret".Sha256()) },

                    Scopes = { "api" }
                },
            };

        public static IEnumerable<Client> Clients =>
            new Client[]
            {
                new Client
                {
                    ClientId = "interactive.confidential",
                    ClientName = "Interactive client (Code with PKCE)",

                    //RedirectUris = { "https://notused" },
                    //PostLogoutRedirectUris = { "https://notused" },
                    RedirectUris = { "https://localhost:5010/signin-oidc" },
                    BackChannelLogoutUri = "https://localhost:5010/bff/backchannel",
                    PostLogoutRedirectUris = { "https://localhost:5010/signout-callback-oidc" },

                    ClientSecrets = { new Secret("secret".Sha256()) },

                    AllowedGrantTypes = GrantTypes.CodeAndClientCredentials,
                    AllowedScopes = AllScopes,

                    AllowOfflineAccess = true,
                    RefreshTokenUsage = TokenUsage.ReUse,
                    RefreshTokenExpiration = TokenExpiration.Sliding
                },
            };
```

I used the IdentityServer Configuration form you gitHub repository. Things I noticed:
1. In IdentityServer Config:
`                    //RedirectUris = { "https://notused" },
                    //PostLogoutRedirectUris = { "https://notused" },`do not work (which sounds logical to me after reading your documentation) So I replaced the values with the localhost addresses from the FrontendHost. 
2. In your repository, you have the following configuration  
`            // registers HTTP client that uses the managed user access token
            services.AddUserAccessTokenHttpClient("api_client", configureClient: client =>
            {
                client.BaseAddress = new Uri("https://localhost:5002/");
            });`
I was wondering as no project from the solution ever uses this localhost address. What is the purpose of declaring it that way? 

I would highly appreciate your help in setting up my Idp pointing me in the corect direction. Thank you!

**Comment by SethNafet** [comment]
I git it to work now. I am just wondering if the solution is intentional. the URI declaration `options.Authority = "https://localhost:5001";` works for the frontent but not for the backendApiHost. Only after changing to 'options.Authority = "https://localhost:5001/";' I can make the full roundtrip. Is this intended?

**Comment by brockallen** [comment]
I didn't look into this yet, but you're saying the trailing slash was the difference between making it work an not?

**Comment by brockallen** [comment]
> Eventually it worked as intended. I then tested without the trailing slash (wich was my last edit) and it kept working. I guess I run into some timeout behavoir earlier.

Ok, yes -- it should work both ways (thus my immediate question). 

`AddUserAccessTokenHttpClient` adds a named HttpClient to DI and configures "globally" it as desired -- in the sample, it sets the base address for all outbound HTTP calls. All this is doing is using the HTTP client factory plumbing from ASP.NET Core and it really a helper from this library: https://identitymodel.readthedocs.io/en/latest/aspnetcore/web.html

**Comment by SethNafet** [comment]
Dear Allen, thank you for the explanation. I am not sure if I understand it well. As from the documentation, I could define userClaims in IdentityResources, ApiScopes or ApiResources, or all?. In any case I would need the profile service to manage the userClaims. Now this leaves the question how the remote API can add Claims to the access or identity token. In our case, Users get rights dynamically based on their membership in organisations. This data is not part of the ApplicationUser db. It is part of the db managed by the backendApi. I thought the profile service is implemented olongside IdentityServer? previously i just hocked up a simple webclaimtransformer at the Api and it did the job. How will IdentityServer profile service know what to put into the userClaims?

**Comment by brockallen** [comment]
ApiScope models your APIs, so that's where you control the config for the user claims that go into the access token. And yes, in all cases, you will need a profile service to get the claims from your custom user DB. If you use ASP.NET Identity, we have a basic implementation already done for you. 

If at the remote API you need/want more claims, then look into claims transformation in the API app's pipeline.

**Comment by SethNafet** [comment]
@brockallen thanks you for the support. I have all claims in the appropriate tokens. However, I still need to figure out how to have them in the session cookie. As this is not related to an issue with IdentityServer, I will start a Q&A discussion in the BFF section. I guess that the issue can be closed.
